src/web/http/manager.cpp

```diff
@@ -0,0 +1,1323 @@
+            }
+            else
+            {
+                LOG_VERBOSE(LOG_SERVER_HTTP)
+                    << "Failed to parse HTTP request from " << request;
+                return handle_connection(connection, event::error);
+            }
+
+            break;
+        }
+
+        case event::write:
+        {
+            // Should never get here since writes are handled
+            // elsewhere.
+            BITCOIN_ASSERT(false);
+            return false;
+        }
+
+        case event::error:
+        case event::closing:
+        {
+            if ((connection == nullptr) || connection->closed())
+                return false;
+
+            LOG_VERBOSE(LOG_SERVER_HTTP)
+                << "Connection closing: " << connection;
+            handler_(connection, event::closing, nullptr);
+            remove_connection(connection);
+            connection->close();
+            return false;
+        }
+
+        default:
+            return false;
+    }
+
+    return true;
+}
+
+#ifdef WITH_MBEDTLS
+// static
+int manager::ssl_send(void* data, const unsigned char* buffer, size_t length)
+{
+    auto connection = reinterpret_cast<http::connection*>(data);
+#ifdef MSG_NOSIGNAL
+    int flags = MSG_NOSIGNAL;
+#else
+    int flags = 0;
+#endif
+    const auto sent = static_cast<int>(send(connection->socket(), buffer,
+        length, flags));
+    if (sent >= 0)
+        return sent;
+
+    return ((would_block(sent) || sent == EINPROGRESS) ?
+        MBEDTLS_ERR_SSL_WANT_WRITE : -1);
+}
+
+// static
+int manager::ssl_receive(void* data, unsigned char* buffer, size_t length)
+{
+    auto connection = reinterpret_cast<http::connection*>(data);
+    auto read = static_cast<int>(recv(connection->socket(), buffer, length, 0));
+    if (read >= 0)
+        return read;
+
+    return ((would_block(read) || read == EINPROGRESS) ?
+        MBEDTLS_ERR_SSL_WANT_READ : -1);
+}
+#endif
+
+bool manager::transfer_file_data(connection_ptr& connection)
+{
+    std::array<unsigned char, transfer_buffer_length> buffer{};
+
+    auto& file_transfer = connection->file_transfer();
+    if (!file_transfer.in_progress)
+        return false;
+
+    auto data = reinterpret_cast<void*>(buffer.data());
+    auto amount_to_read = std::min(transfer_buffer_length,
+        file_transfer.length - file_transfer.offset);
+
+    int32_t read = fread(data, sizeof(unsigned char), amount_to_read,
+        file_transfer.descriptor);
+
+    auto success = ((read == amount_to_read) ||
+        ((read < amount_to_read) && feof(file_transfer.descriptor)));
+
+    if (!success)
+    {
+        if (file_transfer.in_progress)
+        {
+            fclose(file_transfer.descriptor);
+            file_transfer.in_progress = false;
+            file_transfer.offset = 0;
+            file_transfer.length = 0;
+        }
+        return false;
+    }
+
+    int32_t written = connection->write(buffer.data(), read);
+    if (written < 0)
```

Comment:
Based on the subsequent log message it appears that this condition should be:

`if (written != read)...`

Reply:
Good find.  I believe that check as originally written is ok (and the log could be adjusted since it is really showing the write failure, not the amount of bytes written), but then the bug is that `file_transfer.offset` should be incremented by `written`, not `read`.  It's not technically required that all of that write complete (or be buffered) at once.

Reply:
On second thought, it may be better to adjust that check as you specified (although the log message could still be adjusted) because otherwise we'd have to account for seeking backwards by the amount difference (read - written) to prevent data corruption in the stream.